comms/src/outbound_message_service/outbound_message_pool/message_retry_service.rs

```diff
@@ -0,0 +1,203 @@
+
+                let outbound_conn = Connection::new(&context, Direction::Outbound)
+                    .set_name("omp-retry-service-outbound")
+                    .establish(&outbound_address)
+                    .map_err(OutboundMessagePoolError::ConnectionError)?;
+
+                loop {
+                    match message_retry_queue.run(&inbound_conn, &outbound_conn) {
+                        Ok(_) => break,
+                        Err(err) => {
+                            error!(target: LOG_TARGET, "Outbound message retry pool errored: {:?}", err);
+                            warn!(
+                                target: LOG_TARGET,
+                                "Restarting outbound message retry queue after failure."
+                            );
+                            thread::sleep(Duration::from_millis(1000));
+                        },
+                    }
+                }
+
+                info!(target: LOG_TARGET, "Message Retry Queue has cleanly shut down");
+
+                Ok(())
+            })
+            .or_else(|err| {
+                error!(
+                    target: LOG_TARGET,
+                    "Unable to start thread for MessageRetryService: {:?}", err
+                );
+                Err(err)
+            })
+            .unwrap()
+    }
+
+    fn new(config: OutboundMessagePoolConfig, shutdown_signal_rx: Receiver<()>) -> Self {
+        Self {
+            queue: BinaryHeap::new(),
+            config,
+            shutdown_signal_rx,
+        }
+    }
+
+    fn run(
+        &mut self,
+        inbound_conn: &EstablishedConnection,
+        outbound_conn: &EstablishedConnection,
+    ) -> Result<(), OutboundMessagePoolError>
+    {
+        loop {
+            if let Some(frames) = connection_try!(inbound_conn.receive(1000)) {
+                let mut msg = Self::deserialize_outbound_message(frames)?;
+                msg.mark_failed_attempt();
+
+                if msg.num_attempts() > self.config.max_retries {
+                    // Discard message
+                    warn!(
+                        target: LOG_TARGET,
+                        "Discarding message for NodeId {}. Max retry attempts ({}) exceeded.",
+                        msg.destination_node_id(),
+                        self.config.max_retries
+                    );
+                    continue;
+                }
+
+                // Add it to the queue for later retry
+                self.queue.push(msg);
+            }
+
+            match self.shutdown_signal_rx.recv_timeout(Duration::from_millis(1)) {
+                Ok(_) => break,
+                Err(RecvTimeoutError::Timeout) => {},
+                Err(RecvTimeoutError::Disconnected) => {
+                    return Err(OutboundMessagePoolError::ControlMessageSenderDisconnected)
+                },
+            }
+
+            if let Some(msg) = self.queue.peek() {
+                if !msg.is_scheduled() {
```

Comment:
what happens if the first message is not scheduled but the one after that is?

Reply:
The ord implementation for OutboundMessage is ordered from closest to 'now' to furthest from now. BinaryHeap orders them when a message is pushed or poped, so the first in line should always be the most recent.